dimod/utilities.py

```diff
@@ -489,3 +493,121 @@ def update(self, *args, **kwargs):
 def get_include():
     """Return the directory with dimod's header files."""
     return os.path.join(os.path.dirname(__file__), 'include')
+
+
+def _astypearrays(arrays, requirements, min_itemsize, allowed_types):
+    # allowed types can only be numeric for now, see comment below
+    # todo: allow unsafe with warning controlled by kwarg?
+
+    # We need to get the dtype, and as far as I can tell the only way to do
+    # it for array-like is to actually cast to a numpy array
+    arrays = [np.asarray(arr) for arr in arrays]
+
+    # get the dtype we can promote to
+    dtype = reduce(np.promote_types, (arr.dtype for arr in arrays))
+
+    if not any(np.issubdtype(dtype, type_) for type_ in allowed_types):
+        # put together an appropriate error message
+        descriptors = []
+        if np.floating in allowed_types:
+            descriptors.append('floating')
+        if np.integer in allowed_types:
+            descriptors.append('integer')
+        elif np.unsignedinteger in allowed_types:
+            if np.signedinteger in allowed_types:
+                descriptors.append('integer')
+            else:
+                descriptors.append('unsigned integer')
+        elif np.signedinteger in allowed_types:
+            descriptors.append('signed integer')
+
+        raise TypeError(
+            "Cannot safely cast arrays to {} (given {})".format(
+                ', '.join(descriptors),
+                ', '.join(arr.dtype.name for arr in arrays)))
+
+    if min_itemsize is not None:
+        if min_itemsize >= 1:
+            size = str(2**int(np.ceil(np.log2(min_itemsize))))
+        else:
+            size = '1'
+        if np.issubdtype(dtype, np.unsignedinteger):
+            kind = 'u'
+        elif np.issubdtype(dtype, np.signedinteger):
+            kind = 'i'
+        elif np.issubdtype(dtype, np.floating):
+            kind = 'f'
+        else:
+            # we could instead read this from the type string, but it's kind of
+            # pandora's box, because there's also structured arrays, complex,
+            # etc. For now, let's just restrict to numeric.
+            raise RuntimeError("unexpected dtype")
+        dtype = np.promote_types(dtype, kind+size)
+
+    arrays = tuple(np.require(arr, dtype=dtype, requirements=requirements)
+                   for arr in arrays)
+
+    if len(arrays) > 1:
+        return arrays
+    else:
+        return arrays[0]
```

Comment:
Convenient, but confusing. Input is always a sequence, but the output might be a single array. And it's not internal, since callers just propagate this.

Reply:
Sure, but leads to the much nicer (IMO) syntax
```
arr0, arr1, arr2 = asintegerarrays(arr0, arr1, arr2)
arr0, arr1 = asintegerarrays(arr0, arr1)
arr0 = asintegerarrays(arr0)
```
alternative would be something like
```
arr0, arr1, arr2 = asintegerarrays([arr0, arr1, arr2])
arr0, arr1 = asintegerarrays([arr0, arr1])
arr0, = asintegerarrays([arr0])
```

I didn't go with that because numpy's `array-like` gets pretty fiddly when given lists of arrays.